File: 0650-2-keys-keyboard/0650-2-keys-keyboard.java
class Solution {
    public int minSteps(int n) {
        int [] dp = new int [n+1];
        for(int i = 2; i< n+1; i++){
            int divisor = gcd(i);
            dp[i] = dp[divisor] + (i/divisor);
        }
        return dp[n];
    }
    public int gcd(int n){
        for (int i = 2; i <= n/2; i++) {
            if (n % i == 0) {
                return n/i;
            }
        }
        return 1;
    }
}

File: 0650-2-keys-keyboard/0650-2-keys-keyboard.rs
impl Solution {
    pub fn min_steps(n: i32) -> i32 {
        if n == 1 {
            return 0;
        }
        let divisor = Self::gcd(n);
        return Self::min_steps(divisor) + n / divisor;
    }

    pub fn gcd(n: i32) -> i32 {
        for i in 2..=n / 2 {
            if n % i == 0 {
                return n / i;
            }
        }
        return 1;
    }
}